ast/compile.go

```diff
@@ -114,6 +152,222 @@ func (c *Compiler) FlattenErrors() string {
 	return fmt.Sprintf("%d errors occurred:\n%s", len(c.Errors), strings.Join(b, "\n"))
 }
 
+func (c *Compiler) checkBuiltinOperators() {
+	for _, m := range c.Modules {
+		for _, r := range m.Rules {
+			for _, expr := range r.Body {
+				ts, ok := expr.Terms.([]*Term)
+				if !ok {
+					continue
+				}
+				operator := ts[0].Value.(Var)
+				if _, ok := BuiltinMap[operator]; !ok {
+					c.err("bad built-in operator in %v: %v", r.Name, operator)
+				}
+			}
+		}
+	}
+}
+
+// checkRecursion ensures that there are no recursive rule definitions, i.e., there are
+// no cycles in the RuleGraph.
+func (c *Compiler) checkRecursion() {
+	for r := range c.RuleGraph {
+		t := &ruleGraphTraveral{
+			graph:   c.RuleGraph,
+			visited: map[*Rule]struct{}{},
+		}
+		if p := util.DFS(t, r, r); len(p) > 0 {
+			n := []string{}
+			for _, x := range p {
+				n = append(n, string(x.(*Rule).Name))
+			}
+			c.err("recursion found in %v: %v", r.Name, strings.Join(n, ", "))
+		}
+	}
+}
+
+// checkSafetyBody ensures that variables appearing in negated expressions or non-target
+// positions of built-in expressions will be bound when evaluating the rule from left
+// to right. If the variable is bound in an expression that would be processed *after*
+// the negated or built-in expression, the expressions will be re-ordered.
+func (c *Compiler) checkSafetyBody() {
```

Comment:
This is a beast of a function.  

Here's the safety algorithm in my head.  It (i) takes one pass over the rule, (ii) minimally reorders the rule body to achieve safety, and (iii) computes exactly the elements of the body that are unsafe, if any.

Walk over the body of the rule and as we do keep a list of safe variables and a list of unsafe Body elements.  
- When we hit a negated element, we leave it where it is if all of its variables are in the current list of safe variables.  Otherwise we add remove it from the body and add it to the list of unsafe Body elements.
- When we hit a builtin, we check if all its inputs are in the list of safe variables and if so leave it where it is, adding all its output variables to the safe variables list.  Otherwise, we remove it from the Body and add it to the list of unsafe Body elements.
- When we hit a positive, non-builtin element, we add all the variables to our list of safe variables.  Then we move all of those elements from the unsafe Body list to the rule Body that either (i) the element is a builtin and all its inputs are now safe or (ii) the element is negated and all its inputs are safe. 

If the rule is unsafe, the unsafe Body list will contain exactly those Body elements that are problematic.